### PR TITLE
[FSDP][1/N] Add `_get_fsdp_states()`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #90959 [FSDP][6/N] Add note explaining idioms for `_FSDPState` traversal
* #90874 [FSDP][5/N] Add manual "wrapping" support for `fully_shard`
* #90871 [FSDP][4/N] Refactor func to share state/init handle attrs
* #90862 [FSDP][3/N] Move `fsdp_modules(root_only=True)` -> `_get_fsdp_root_states()`
* #90861 [FSDP][2/N] Move `fsdp_modules(root_only=False)` -> `_get_fsdp_states()`
* #90864 [FSDP][Easy] Rename `entry` -> `fsdp_module` to be more descriptive
* **#90860 [FSDP][1/N] Add `_get_fsdp_states()`**

- This PR introduces `_get_fsdp_states(module: nn.Module) -> List[_FSDPState]` to prepare for `fully_shard` manual "wrapping".
    - ~~I place it in `_runtime_utils.py`, not `_common_utils.py`, because in a follow-up PR, I will add `_get_root_fsdp_states()`, which requires `_lazy_init()`. I concluded that it would be preferred to have both of these getters be in the same place than to have them split, even if that means that `_get_fsdp_states()` is in `_runtime_utils.py`.~~ Due to circular import issues, I think I should still put it in `_common_utils.py`.
- This PR changes `FullyShardedDataParallel.fsdp_modules()` to be backed by `_get_fsdp_states()`.

### CODE REVIEW

File: torch/distributed/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -1249,3 +1260,26 @@ def _cast_buffers_to_dtype_and_device(
...
+
+    NOTE: For now, we must pass an ``nn.Module`` as the argument because
+    ``_FSDPState`` does not support graph traversal.
+    """
+    fsdp_states: List[_FSDPState] = []
+    visited_fsdp_states: Set[_FSDPState] = set()
+    for submodule in module.modules():
+        optional_state = _get_module_fsdp_state(submodule)
+        if optional_state is not None and optional_state not in visited_fsdp_states:
+            visited_fsdp_states.add(optional_state)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/_runtime_utils.py:1283]: do we need visited set if we just let fsdp_states be a set and convert back to a list?

**Review by rohan-varma — APPROVED** [review]
Would it be useful to have a unittest?

File: torch/distributed/fsdp/_runtime_utils.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/_runtime_utils.py:1283]: We need to preserve the ordering. Let me make that clear.

**Comment by awgu** [comment]
> Would it be useful to have a unittest?

I think for this particular PR in the stack, it is okay to not include a unit test, since we just nee to make sure `fsdp_modules` backward compatibility is preserved. For later PRs in the stack, we will naturally require that this function works in order for the composable path to work.

File: torch/distributed/fsdp/_common_utils.py [hunk 2]
```diff
@@ -40,6 +40,29 @@ def _get_module_fsdp_state(module: nn.Module) -> Optional[_FSDPState]:
...
+    NOTE: For now, we must pass an ``nn.Module`` as the argument because
+    ``_FSDPState`` does not support graph traversal.
+    """
+    fsdp_states: List[_FSDPState] = []
+    visited_fsdp_states: Set[_FSDPState] = set()
+    for submodule in module.modules():
+        optional_state = _get_module_fsdp_state(submodule)
+        if optional_state is not None and optional_state not in visited_fsdp_states:
+            visited_fsdp_states.add(optional_state)
+            fsdp_states.append(optional_state)
```
Comment (mrshenli) [inline torch/distributed/fsdp/_common_utils.py:62]: do we care about the order here (curious as we are using set + list)?

Comment (awgu) [reply torch/distributed/fsdp/_common_utils.py:62]: Yes. I want to preserve the depth-first order that `module.modules()` follows, which is why I have this function return a list (and only use the set for tracking visited-ness).

**Pull request closed without merging**